src/HexagonOptimize.cpp

```diff
@@ -0,0 +1,1117 @@
+        // does not deinterleave, and then opportunistically select
+        // the interleaving alternative when we can cancel out to the
+        // interleave.
+        struct DeinterleavingAlternative {
+            string name;
+            vector<Expr> extra_args;
+        };
+        static std::map<string, DeinterleavingAlternative> deinterleaving_alts = {
+            { "halide.hexagon.pack.vh", { "halide.hexagon.trunc.vh" } },
+            { "halide.hexagon.pack.vw", { "halide.hexagon.trunc.vw" } },
+            { "halide.hexagon.pack_satub.vh", { "halide.hexagon.trunc_satub.vh" } },
+            { "halide.hexagon.pack_sath.vw", { "halide.hexagon.trunc_sath.vw" } },
+            // For this one, we don't have a simple alternative. But,
+            // we have a shift-saturate-narrow that we can use with a
+            // shift of 0.
+            { "halide.hexagon.pack_satuh.vw", { "halide.hexagon.trunc_satuh_shr.vw.w", { 0 } } },
+        };
+
+        if (is_native_deinterleave(op) && yields_interleave(args[0])) {
+            // This is a deinterleave of an interleave! Remove them both.
+            expr = remove_interleave(args[0]);
+        } else if (is_interleavable(op) && yields_removable_interleave(args)) {
+            // All the arguments yield interleaves (and one of
+            // them is an interleave), create a new call with the
+            // interleave removed from the arguments.
+            for (Expr &i : args) {
+                i = remove_interleave(i);
+            }
+            expr = Call::make(op->type, op->name, args, op->call_type,
+                              op->func, op->value_index, op->image, op->param);
+            // Add the interleave back to the result of the call.
+            expr = native_interleave(expr);
+        } else if (deinterleaving_alts.find(op->name) != deinterleaving_alts.end() &&
+                   yields_removable_interleave(args)) {
+            // This call has a deinterleaving alternative, and the
+            // arguments are interleaved, so we should use the
+            // alternative instead.
+            const DeinterleavingAlternative &alt = deinterleaving_alts[op->name];
+            for (Expr &i : args) {
+                i = remove_interleave(i);
+            }
+            for (Expr i : alt.extra_args) {
+                args.push_back(i);
+            }
+            expr = Call::make(op->type, alt.name, args, op->call_type);
+        } else if (changed) {
+            expr = Call::make(op->type, op->name, args, op->call_type,
+                              op->func, op->value_index, op->image, op->param);
+        } else {
+            expr = op;
+        }
+    }
+
+    using IRMutator::visit;
+};
+
+// Find a conservative upper bound of an expression.
+class UpperBound : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const Sub *op) {
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
```

Comment:
This doesn't look right. If you have a-b, and you want an upper bound, then you need the *lower* bound of b.